### PR TITLE
P2322R6 ranges::fold

### LINKED ISSUE #5604: [2022-07 LWG Motion 17] P2322R6 ranges::fold
[P2322R6](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2322r6.html)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #5604
Fixes cplusplus/papers#1004

### CODE REVIEW

File: source/algorithms.tex [hunk 1]
```diff
@@ -1169,6 +1172,91 @@
...
+      @\exposconcept{indirectly-binary-right-foldable}@<T, I> F>
+      constexpr auto fold_right(I first, S last, T init, F f);
+
+    template<@\libconcept{bidirectional_range}@ R, class T,
+      @\exposconcept{indirectly-binary-right-foldable}@<T, iterator_t<R>> F>
+      constexpr auto fold_right(R&& r, T init, F f);
+
+    template <@\libconcept{bidirectional_iterator}@ I, @\libconcept{sentinel_for}@<I> S,
+      indirectly-binary-right-foldable<iter_value_t<I>, I> F>
+      requires constructible_from<iter_value_t<I>, iter_reference_t<I>>
```
Comment (JohelEGP) [inline source/algorithms.tex:1231]:
```suggestion
      @\exposconcept{indirectly-binary-right-foldable}@<iter_value_t<I>, I> F>
      requires @\libconcept{constructible_from}@<iter_value_t<I>, iter_reference_t<I>>
```

File: source/algorithms.tex [hunk 2]
```diff
@@ -1169,6 +1172,91 @@
...
+
+    template<@\libconcept{input_range}@ R, class T, @\exposconcept{indirectly-binary-left-foldable}@<T, iterator_t<R>> F>
+      constexpr auto fold_left(R&& r, T init, F f);
+
+    template<@\libconcept{input_iterator}@ I, @\libconcept{sentinel_for}@<I> S,
+             @\exposconcept{indirectly-binary-left-foldable}@<iter_value_t<I>, I> F>
+      requires @\libconcept{constructible_from}@<iter_value_t<I>, iter_reference_t<I>>
+      constexpr auto fold_left_first(I first, S last, F f);
+
+    template <@\libconcept{input_range}@ R, @\exposconcept{indirectly-binary-left-foldable}@<range_value_t<R>, iterator_t<R>> F>
```
Comment (JohelEGP) [inline source/algorithms.tex:1217]: Replace all `template <` with `template<`.

File: source/algorithms.tex [hunk 3]
```diff
@@ -4384,6 +4490,170 @@
...
+return optional<U>(in_place,
+  ranges::fold_right(std::move(first), tail, iter_value_t<I>(*tail), std::move(f)));
+\end{codeblock}
+\end{itemdescr}
+
+\indexlibraryglobal{fold_left_with_iter}%
+\begin{itemdecl}
+template<@\libconcept{input_iterator}@ I, @\libconcept{sentinel_for}@<I> S, class T,
+         @\exposconcept{indirectly-binary-left-foldable}@<T, I> F>
+  constexpr @\seebelow@ fold_left_with_iter(I first, S last, T init, F f);
```
Comment (JohelEGP) [inline source/algorithms.tex:4589]:
```suggestion
  constexpr @\seebelow@ ranges::fold_left_with_iter(I first, S last, T init, F f);
```

File: source/algorithms.tex [hunk 4]
```diff
@@ -4384,6 +4490,170 @@
...
+template<@\libconcept{input_range}@ R, class T, @\exposconcept{indirectly-binary-left-foldable}@<T, iterator_t<R>> F>
+  constexpr @\seebelow@ fold_left_with_iter(R&& r, T init, F f);
```
Comment (JohelEGP) [inline source/algorithms.tex:4591]:
```suggestion
  constexpr @\seebelow@ ranges::fold_left_with_iter(R&& r, T init, F f);
```

File: source/algorithms.tex [hunk 5]
```diff
@@ -4384,6 +4490,170 @@
...
+\tcode{fold_left_with_iter_result<borrowed_iterator_t<R>, U>}
+for the second overload.
+\end{itemdescr}
+
+\indexlibraryglobal{fold_left_first_with_iter}%
+\begin{itemdecl}
+template<@\libconcept{input_iterator}@ I, @\libconcept{sentinel_for}@<I> S,
+         @\exposconcept{indirectly-binary-left-foldable}@<iter_value_t<I>, I> F>
+  requires @\libconcept{constructible_from}@<iter_value_t<I>, iter_reference_t<I>>
+  constexpr @\seebelow@ fold_left_first_with_iter(I first, S last, F f)
```
Comment (JohelEGP) [inline source/algorithms.tex:4623]:
```suggestion
  constexpr @\seebelow@ ranges::fold_left_first_with_iter(I first, S last, F f);
```

File: source/algorithms.tex [hunk 6]
```diff
@@ -4384,6 +4490,170 @@
...
+template<@\libconcept{input_range}@ R, @\exposconcept{indirectly-binary-left-foldable}@<range_value_t<R>, iterator_t<R>> F>
+  requires @\libconcept{constructible_from}@<range_value_t<R>, range_reference_t<R>>
+  constexpr @\seebelow@ fold_left_first_with_iter(R&& r, F f);
```
Comment (JohelEGP) [inline source/algorithms.tex:4626]:
```suggestion
  constexpr @\seebelow@ ranges::fold_left_first_with_iter(R&& r, F f);
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/algorithms.tex (hunk 2 above)
Comment (jensmaurer) [reply source/algorithms.tex]: fixed

File: source/algorithms.tex [hunk 7]
```diff
@@ -1169,6 +1172,97 @@
...
+
+    template<@\libconcept{bidirectional_iterator}@ I, @\libconcept{sentinel_for}@<I> S, class T,
+      @\exposconcept{indirectly-binary-right-foldable}@<T, I> F>
+      constexpr auto fold_right(I first, S last, T init, F f);
+
+    template<@\libconcept{bidirectional_range}@ R, class T,
+      @\exposconcept{indirectly-binary-right-foldable}@<T, iterator_t<R>> F>
+      constexpr auto fold_right(R&& r, T init, F f);
+
+<<<<<<< Updated upstream
```
Comment (JohelEGP) [inline source/algorithms.tex:1229]:
```suggestion
```

File: source/algorithms.tex [hunk 8]
```diff
@@ -1169,6 +1172,97 @@
...
+    template <@\libconcept{bidirectional_iterator}@ I, @\libconcept{sentinel_for}@<I> S,
+      @\exposconcept{indirectly-binary-right-foldable}@<iter_value_t<I>, I> F>
+      requires @\libconcept{constructible_from}@<iter_value_t<I>, iter_reference_t<I>>
+=======
+    template<@\libconcept{bidirectional_iterator}@ I, @\libconcept{sentinel_for}@<I> S,
+      indirectly-binary-right-foldable<iter_value_t<I>, I> F>
+      requires constructible_from<iter_value_t<I>, iter_reference_t<I>>
+>>>>>>> Stashed changes
```
Comment (JohelEGP) [inline source/algorithms.tex:1237]:
```suggestion
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

**Pull request merged**